pallets/api/src/extension.rs

```diff
@@ -0,0 +1,131 @@
+		log::debug!(target: Self::LOG_TARGET, "prepender: version={version}, module={module}, index={index}");
+		value
+	}
+}
+
+/// Matches on the first byte of a function identifier only.
+pub struct IdentifiedByFirstByteOfFunctionId<T>(PhantomData<T>);
+impl<T: Get<u8>> Matches for IdentifiedByFirstByteOfFunctionId<T> {
+	fn matches(env: &impl Environment) -> bool {
+		func_id(env) == T::get()
+	}
+}
+
+/// A log target for dispatched calls.
+pub struct DispatchCallLogTarget;
+impl LogTarget for DispatchCallLogTarget {
+	const LOG_TARGET: &'static str = "pop-api::extension::dispatch";
+}
+
+/// A log target for state reads.
+pub struct ReadStateLogTarget;
+impl LogTarget for ReadStateLogTarget {
+	const LOG_TARGET: &'static str = "pop-api::extension::read-state";
+}
+
+/// Conversion of a `DispatchError` to a versioned error.
+pub struct VersionedErrorConverter<E>(PhantomData<E>);
+impl<Error: From<(DispatchError, u8)> + Into<u32> + Debug> pop_chain_extension::ErrorConverter
+	for VersionedErrorConverter<Error>
+{
+	/// The log target.
+	const LOG_TARGET: &'static str = "pop-api::extension::converters::versioned-error";
+
+	/// Converts the provided error.
+	///
+	/// # Parameters
+	/// - `error` - The error to be converted.
+	/// - `env` - The current execution environment.
+	fn convert(error: DispatchError, env: &impl Environment) -> Result<RetVal> {
+		// Defer to supplied versioned error conversion type
+		let version = version(env);
+		log::debug!(target: Self::LOG_TARGET, "versioned error converter: error={error:?}, version={version}");
+		let error: Error = (error, version).into();
+		log::debug!(target: Self::LOG_TARGET, "versioned error converter: converted error={error:?}");
+		Ok(RetVal::Converging(error.into()))
+	}
+}
+
+/// Conversion of a read result to a versioned read result.
+pub struct VersionedResultConverter<S, T>(PhantomData<(S, T)>);
+impl<Source: Debug, Target: From<(Source, u8)> + Debug> Converter
+	for VersionedResultConverter<Source, Target>
+{
+	/// The type of value to be converted.
+	type Source = Source;
+	/// The target type.
+	type Target = Target;
+	/// The log target.
+	const LOG_TARGET: &'static str = "pop-api::extension::converters::versioned-result";
+
+	/// Converts the provided value.
+	///
+	/// # Parameters
+	/// - `value` - The value to be converted.
+	/// - `env` - The current execution environment.
+	fn convert(value: Self::Source, env: &impl Environment) -> Self::Target {
+		// Defer to supplied versioned result conversion type
+		let version = version(env);
+		log::debug!(target: Self::LOG_TARGET, "versioned result converter: result={value:?}, version={version}");
+		let converted: Target = (value, version).into();
+		log::debug!(target: Self::LOG_TARGET, "versioned result converter: converted result={converted:?}");
+		converted
+	}
+}
+
+fn func_id(env: &impl Environment) -> u8 {
+	// TODO: update once the encoding scheme order has been finalised: expected to be env.ext_id().to_le_bytes()[0]
```

Comment:
do we want to merge with this TODO being unresolved?

Reply:
As mentioned separately, it was agreed that this would be dealt with separately. I dont believe much has been finalised there yet.